### PR TITLE
fixed bucket name validation

### PR DESCRIPTION
**Pull request by c2thorn**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/12226

Rewrote validation logic and added much-needed unit testing. Invalid cases were tested against the API

[Requirements ](https://cloud.google.com/storage/docs/naming-buckets) implemented:

- Bucket names can only contain lowercase letters, numeric characters, dashes (-), underscores (_), and dots (.). Spaces are not allowed.
- Bucket names must start and end with a number or letter.
- Bucket names must contain 3-63 characters. Names containing dots can contain up to 222 characters, but each dot-separated component can be no longer than 63 characters.
- Bucket names cannot begin with the "goog" prefix.
- Bucket names cannot contain "google" 

NOT implemented:
- Names containing dots require [verification](https://cloud.google.com/storage/docs/domain-name-verification).
- Bucket names cannot be represented as an IP address in dotted-decimal notation (for example, 192.168.5.4).
- Bucket names cannot contain close misspellings of "google", such as "g00gle".

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
storage: fixed a bug in `google_storage_bucket` where `name` was incorrectly validated.
```

### CODE REVIEW

**Comment by c2thorn** [comment]
@prateek2408 fyi

**Comment by c2thorn** [comment]
Current running VCR tests are not related

File: mmv1/third_party/terraform/utils/utils.go [hunk 1]
```diff
@@ -527,26 +527,26 @@ func fake404(reasonResourceType, resourceName string) *googleapi.Error {
...
+		return fmt.Errorf("error: bucket name %s cannot start with %q", name, "goog")
+	}
+
+	if strings.Contains(name, "google") {
+		return fmt.Errorf("error: bucket name %s cannot contain %q", name, "google")
+	}
+
+	valid, _ := regexp.MatchString("^[a-z0-9][a-z0-9_.-]{1,220}[a-z0-9]$", name)
+	if !valid {
+		return fmt.Errorf("error: bucket name validation failed %v", name)
```
Comment (daniel-cit) [inline mmv1/third_party/terraform/utils/utils.go:543]:
```suggestion
		return fmt.Errorf("error: bucket name validation failed %v. See https://cloud.google.com/storage/docs/naming-buckets", name)
```

would a link to the documentation be helpful?

Comment (c2thorn) [reply mmv1/third_party/terraform/utils/utils.go:543]: @daniel-cit not a bad idea

**Comment by c2thorn** [comment]
I'll wait for `tpg-test` step in `generate-diffs` to pass before merging

**Comment by hdotking** [comment]
Which terraform google provider version/setup do I need to be using to make use of this bug fix?

**Comment by c2thorn** [comment]
This will be a part of next week's release (v4.32) @hdotking

**Pull request merged**